cms/models/placeholdermodel.py

```diff
@@ -739,3 +739,76 @@ def _recalculate_plugin_positions(self, language):
             raise RuntimeError(
                 '{} is not supported by django-cms'.format(connection.vendor)
             )
+
+    def check_tree(self, language=None):
+        """checks the plugin tree if the placeholder for common inconsistencies and returns a list
+        of messages describing the inconsistency. Returns list of messages."""
+        messages = []
+
+        if language is None:
+            languages = self.cmsplugin_set.order_by('language').values_list('language', flat=True).distinct()
+            for language in languages:
+                message = self.check_tree(language)
+                if message is not None:
+                    messages += message
+            return messages
+
+        # Check 1: Positions are consecutive starting at 1
+        position_list = list(self.cmsplugin_set.values_list('position', flat=True))
+        if position_list != list(range(1, self.get_last_plugin_position(language)+1)):
+            messages.append(f"{language}: Non consecutive position entries: {position_list}")
+
+        # Check 2: Children AFTER parents
+        parent_list = list(self.cmsplugin_set.values_list('parent', flat=True))
+        for parent in parent_list:
+            if parent is not None:
+                parent_position = self.cmsplugin_set.filter(pk=parent).first()
+                if parent_position is not None:
+                    children_positions = self.cmsplugin_set.filter(parent=parent).values_list('position', flat=True)
+                    if min(children_positions) <= parent_position.position:
+                        messages.append(f"{language}: Children with positions lower than their parent's (id={parent}) position")
+
+        # Check 3: parents belonging to other placeholders
+        for plugin in self.cmsplugin_set.all():
+            if plugin.parent and plugin.parent.placeholder != self:
+                messages.append(f"{language}: Plugins claim to be children of parents in a different placeholder")
+
+        return messages
+
+    def fix_tree(self, language=None):
```

Comment:
I fear that this will just end up like django-cms v3 where production container spin up always runs this command. It's not very efficient and people with versioning will want this to work very differently.

Reply:
So far this is only a method for the Placeholder model. Not a management command. If added as a management command it would need to be only invoked on placeholders in draft models. If we fix the plugin trees we might never need it as a management command. Right now, @marksweb at least will need the fix for his site to work.

I suggest to add a management command for the develop-4. branch. We can decide to remove it before publishing v4. Clearly it should not be needed in a published version.

Reply:
The design principle of the core is to be as simple as possible, temporary logic should be a 3rd party package, not part of the core. It also allows changes to be optional and provides a much faster turn around. 

I agree Mark needs a fast solution, this can be a management command in the djangocms-frontend package because it is this package that is seeing these issues. The only time I saw the same issue is with the text editor logic I keep referring to. It had exactly the same side affects and it was because the cms wasn't used to get the position. Once the text editor got the position properly no position issues were seen again. 

```
        last_position = placeholder.get_last_plugin_position(language) or 0
        # A shift is only needed if the distance between the new plugin
        # and the last plugin is greater than 1 position.
        needs_shift = (position - last_position) < 1

        if needs_shift:
            # shift to the right to make space for the new ghost plugin
            placeholder._shift_plugin_positions(
                language,
                start=position,
                offset=last_position,
            )
```

Reply:
The fix tre command you are creating probably only needs to be this method:
```
placeholder._shift_plugin_positions(
    language,
    start=position,
    offset=last_position,
)
```

Reply:
@marksweb could you try this script locally and see if it reorders the plugins to get you out of your blocker? ^^

Reply:
You can use version compare manually to see if it has reordered anything unintentionally. Compare is useless if all versions have been manipulated.

Reply:
For me it's also fine to have a fix-tree management command in a separate mini app or even script. Thanks for the code snippet to get the draft pages. I was thinking of let it run over all placeholders that are assigned to a ContentType object with "draft" status in versioning and, e.g., the clipboard.

Reply:
@Aiky30 Yeah need to give it a run. That earlier version that just went over all placeholders did resolve it, but obviously wasn't the correct approach so need to run this now.

Reply:
@Aiky30 I've tried to run your example and it didn't change what's happening.

Reply:
Sorry guys, I saw what u thought was this just being merged without any approvals hence why I threw my toys out of the pram.

I do have genuine concerns at the scale if the changes here. 

If you're blocked then I guess you have to do something. 

@fsbraun  The tests will need to create multiple trees in different languages. 

Does there need to be the amount if changes made to the tree or can this be made less intrusive?

@marksweb it's a shame that the script I sent didn't work, I imagine it could with some tweaking but I'm away all weekend so not going to be much help on getting something fixed fast.

I would still lean towards making sure any fixing of previous trees is handled outside of the cms, this is very hacky and I'm pretty sure will crash on large projects.